File: platforms/spigot/src/main/java/com/cjcrafter/foliascheduler/bukkit/BukkitEntityScheduler.java
package com.cjcrafter.foliascheduler.bukkit;

import com.cjcrafter.foliascheduler.EntitySchedulerImplementation;
import com.cjcrafter.foliascheduler.TaskImplementation;
import org.bukkit.entity.Entity;
import org.bukkit.plugin.Plugin;
import org.bukkit.scheduler.BukkitRunnable;
import org.jetbrains.annotations.ApiStatus;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.function.Function;

@ApiStatus.Internal
public class BukkitEntityScheduler implements EntitySchedulerImplementation {

    private final @NotNull Plugin plugin;
    private final @NotNull Entity entity;

    @ApiStatus.Internal
    public BukkitEntityScheduler(@NotNull Plugin plugin, @NotNull Entity entity) {
        this.plugin = plugin;
        this.entity = entity;
    }

    private <T> @NotNull BukkitRunnable buildBukkitRunnable(
        @NotNull Function<TaskImplementation<T>, T> function,
        @NotNull BukkitTask<T> taskImplementation
    ) {
        return new BukkitRunnable() {
            @Override
            public void run() {
                taskImplementation.setCallback(function.apply(taskImplementation));
                taskImplementation.asFuture().complete(taskImplementation);
            }
        };
    }

    private boolean isRetired(@Nullable Runnable retired) {
        if (!entity.isValid()) {
            if (retired != null)
                retired.run();
            return true;
        }
        return false;
    }

    @Override
    public boolean execute(@NotNull Runnable run, @Nullable Runnable retired, long delay) {
        if (isRetired(retired))
            return false;

        new BukkitRunnable() {
            @Override
            public void run() {
                if (isRetired(retired))
                    return;
                run.run();
            }
        }.runTaskLater(plugin, delay);
        return true;
    }

    @Override
    public @Nullable <T> TaskImplementation<T> run(@NotNull Function<TaskImplementation<T>, T> function, @Nullable Runnable retired) {
        if (isRetired(retired))
            return null;

        BukkitTask<T> taskImplementation = new BukkitTask<>(plugin, false);
        BukkitRunnable runnable = buildBukkitRunnable(function, taskImplementation);
        taskImplementation.setScheduledTask(runnable.runTask(plugin));
        return taskImplementation;
    }

    @Override
    public @Nullable <T> TaskImplementation<T> runDelayed(@NotNull Function<TaskImplementation<T>, T> function, @Nullable Runnable retired, long delay) {
        if (isRetired(retired))
            return null;

        BukkitTask<T> taskImplementation = new BukkitTask<>(plugin, false);
        BukkitRunnable runnable = buildBukkitRunnable(function, taskImplementation);
        taskImplementation.setScheduledTask(runnable.runTaskLater(plugin, delay));
        return taskImplementation;
    }

    @Override
    public @Nullable <T> TaskImplementation<T> runAtFixedRate(@NotNull Function<TaskImplementation<T>, T> function, @Nullable Runnable retired, long delay, long period) {
        if (isRetired(retired))
            return null;

        BukkitTask<T> taskImplementation = new BukkitTask<>(plugin, true);
        BukkitRunnable runnable = buildBukkitRunnable(function, taskImplementation);
        taskImplementation.setScheduledTask(runnable.runTaskTimer(plugin, delay, period));
        return taskImplementation;
    }
}

File: src/main/kotlin/com/cjcrafter/scheduler/bukkit/BukkitEntityScheduler.kt
package com.cjcrafter.scheduler.bukkit

import com.cjcrafter.scheduler.EntitySchedulerImplementation
import com.cjcrafter.scheduler.TaskImplementation
import org.bukkit.entity.Entity
import org.bukkit.plugin.Plugin
import org.bukkit.scheduler.BukkitRunnable
import java.util.function.Consumer

class BukkitEntityScheduler(
    private val plugin: Plugin,
    private val entity: Entity,
): EntitySchedulerImplementation {
    override fun execute(run: Runnable, retired: Runnable?, delay: Long): Boolean {
        if (!entity.isValid) {
            retired?.run()
            return false
        }

        object : BukkitRunnable() {
            override fun run() {
                if (!entity.isValid) {
                    retired?.run()
                    return
                }
                run.run()
            }
        }.runTask(plugin)
        return true
    }

    override fun run(consumer: Consumer<TaskImplementation>, retired: Runnable?): TaskImplementation? {
        if (!entity.isValid) {
            retired?.run()
            return null
        }

        val taskImplementation = BukkitTask(plugin, false)
        val scheduledTask = object : BukkitRunnable() {
            override fun run() {
                if (!entity.isValid) {
                    retired?.run()
                    return
                }
                consumer.accept(taskImplementation)
                taskImplementation.asFuture().complete(taskImplementation)
            }
        }.runTask(plugin)

        taskImplementation.scheduledTask = scheduledTask
        return taskImplementation
    }

    override fun runDelayed(
        consumer: Consumer<TaskImplementation>,
        retired: Runnable?,
        delay: Long
    ): TaskImplementation? {
        if (!entity.isValid) {
            retired?.run()
            return null
        }

        val taskImplementation = BukkitTask(plugin, false)
        val scheduledTask = object : BukkitRunnable() {
            override fun run() {
                if (!entity.isValid) {
                    retired?.run()
                    return
                }
                consumer.accept(taskImplementation)
                taskImplementation.asFuture().complete(taskImplementation)
            }
        }.runTaskLater(plugin, delay)

        taskImplementation.scheduledTask = scheduledTask
        return taskImplementation
    }

    override fun runAtFixedRate(
        consumer: Consumer<TaskImplementation>,
        retired: Runnable?,
        delay: Long,
        period: Long
    ): TaskImplementation? {
        if (!entity.isValid) {
            retired?.run()
            return null
        }

        val taskImplementation = BukkitTask(plugin, true)
        val scheduledTask = object : BukkitRunnable() {
            override fun run() {
                if (!entity.isValid) {
                    retired?.run()
                    return
                }
                consumer.accept(taskImplementation)
                taskImplementation.asFuture().complete(taskImplementation)
            }
        }.runTaskTimer(plugin, delay, period)

        taskImplementation.scheduledTask = scheduledTask
        return taskImplementation
    }
}
